### PR TITLE
Feat: OPX3.1-dev1

### PR DESCRIPTION
**Pull request by GarrickHe**
* Feature: Add interface name and ifindex to the output list when querying for stats
* Feature: Adding QSA support for 1G SFP media
* Feature: VxLAN Support
* Bugfix: Altered tap fd lock access to be called only after event add/del.
* Bugfix: Used the right mgmt VRF name macro to skip the VRF object creation in the SAI
          for mgmt VRF.
* Bugfix: Avoid sending multiple set requests for speed to PAS
* Bugfix: Sending sequential hwport offset value from port group object to PAS
          for setting media transciever
* Bugfix: Handle the mode change for lag blocked port in netlink handler
* Bugfix: Get the port mode before adding port to LAG in kernel
* Bugfix: Remove the publish of oper state just after connecting interface to NPU port
* Bugfix: Using steady clock for providing independent timestamps
* Bugfix: Unable to delete/clear L2MC on removal of port channel as VLAN member
* Bugfix: MAC learn disable issue in vlan
* Bugfix: The ability to set mac_lear_mode from CPS set for VXLAN
* Update: Exclude vlan dir from compilation
* Update: Enabled default libevent lock mechanism for multi thread access of event base.
* Update: Setting transceiver enable/disable in PAS based on Admin State and not Oper State
* Update: Adding port-group breakout mode for ethernet qsfp28 ports
* Update: Allow ports to be added/removed from VLAN 1 in scaled vlan mode
* Update: Re-apply admin state after membership changes
* Update: Consolidate remote endpoint event into a single event with tunnnel ID
* Update: Allow ports to be added/removed from VLAN 1 in scaled vlan mode
* Update: Use separate service for mac address allocation
* Update: Avoid sending multiple set requests for speed to PAS
* Update: Set only ingress split horizon for pv subport
* Update: Split horizon changes
* Update: MAC learn mode for physical port, LAG
* Update: set_speed method changed to support QSA adaptor
* Update: Check profile type as DDQSFP28 for lane admin up
* Update: Publish vlan sub interface event for regular vlan interfaces as well
* Update: Support to do get on VXLAN type interface
* Update: Send vlan sub interface transaction directly to base
* Update: Correctly handle bridge name when deleting vlan
* Update: Enable/disable interface packet drop status based on its tagged/untagged membership
* Update: Send sequential hwport offset value from port group object to PAS for setting media transciever
* Update: Handle the mode change for lag blocked port in netlink handler
* Update: Get the port mode before adding port to lag in kernel
* Update: Ignore duplex failure while interface creation
* Update: Cache the default vlan id use default vlan id for vxlan in pv mode
* Update: Use vlan id of 1q vlan for untagged port when ports are being migrated to 1d bridge
* Update: Handle tunnel flood disable event and publish an event for the same
* Update: Move untagged vlan to 1d bridge
* Update: VxLAN changes to program npu first and then kernel

Signed-off-by: Garrick He <garrick_he@dell.com>

### CODE REVIEW

File: debian/changelog [hunk 1]
```diff
@@ -1,6 +1,56 @@
+opx-nas-interface (5.20.3) unstable; urgency=medium
+
+  * Feature: Add interface name and ifindex to the output list when querying for stats
+  * Feature: Adding QSA support for 1G SFP media
+  * Feature: VxLAN Support
```
Comment (jeff-yin) [inline debian/changelog]: "Static L2 VxLAN" ?

File: debian/changelog [hunk 2]
```diff
@@ -1,6 +1,56 @@
...
+  * Update: Allow ports to be added/removed from VLAN 1 in scaled vlan mode
+  * Update: Use separate service for mac address allocation
+  * Update: Avoid sending multiple set requests for speed to PAS
+  * Update: Set only ingress split horizon for pv subport
+  * Update: Split horizon changes
+  * Update: MAC learn mode for physical port, LAG
+  * Update: set_speed method changed to support QSA adaptor
+  * Update: Check profile type as DDQSFP28 for lane admin up
+  * Update: Publish vlan sub interface event for regular vlan interfaces as well
+  * Update: Support to do get on VXLAN type interface
```
Comment (jeff-yin) [inline debian/changelog]: "VXLAN" should be "VxLAN"

File: debian/changelog [hunk 3]
```diff
@@ -1,6 +1,56 @@
...
+  * Update: Send vlan sub interface transaction directly to base
```
Comment (jeff-yin) [inline debian/changelog]: What does "base" mean here?

File: debian/changelog [hunk 4]
```diff
@@ -1,6 +1,56 @@
...
+  * Update: Correctly handle bridge name when deleting vlan
+  * Update: Enable/disable interface packet drop status based on its tagged/untagged membership
+  * Update: Send sequential hwport offset value from port group object to PAS for setting media transciever
+  * Update: Handle the mode change for lag blocked port in netlink handler
+  * Update: Get the port mode before adding port to lag in kernel
+  * Update: Ignore duplex failure while interface creation
```
Comment (jeff-yin) [inline debian/changelog]: Suggestion: "while" should be "during"

File: debian/changelog [hunk 5]
```diff
@@ -1,6 +1,56 @@
...
+  * Update: set_speed method changed to support QSA adaptor
+  * Update: Check profile type as DDQSFP28 for lane admin up
+  * Update: Publish vlan sub interface event for regular vlan interfaces as well
+  * Update: Support to do get on VXLAN type interface
+  * Update: Send vlan sub interface transaction directly to base
+  * Update: Correctly handle bridge name when deleting vlan
+  * Update: Enable/disable interface packet drop status based on its tagged/untagged membership
+  * Update: Send sequential hwport offset value from port group object to PAS for setting media transciever
+  * Update: Handle the mode change for lag blocked port in netlink handler
+  * Update: Get the port mode before adding port to lag in kernel
```
Comment (jeff-yin) [inline debian/changelog]: LAG should be capitalized throughout to be consistent with the rest of the CHANGELOG message.

Comment (jeff-yin) [reply debian/changelog]: Same goes for VLAN actually

File: configure.ac [hunk 6]
```diff
@@ -1,7 +1,7 @@
 #                                               -*- Autoconf -*-
 # Process this file with autoconf to produce a configure script.
 AC_PREREQ([2.69])
-AC_INIT([opx-nas-interface], [5.18.0+opx13], [ops-dev@lists.openswitch.net])
+AC_INIT([opx-nas-interface], [5.20.3], [ops-dev@lists.openswitch.net])
```
Comment (atanu-mandal) [inline configure.ac]: Shouldn't be 5.19.0?

File: debian/changelog [hunk 7]
```diff
@@ -1,6 +1,56 @@
+opx-nas-interface (5.20.3) unstable; urgency=medium
```
Comment (atanu-mandal) [inline debian/changelog]: Shouldn't be 5.19.0?

File: scripts/bin/base_nas_fanout_init.sh [hunk 8]
```diff
@@ -0,0 +1,21 @@
+#!/bin/bash
+# Copyright (c) 2015 Dell Inc.
```
Comment (atanu-mandal) [inline scripts/bin/base_nas_fanout_init.sh]: Copyright update?

File: scripts/bin/base_nas_mac_config.py [hunk 9]
```diff
@@ -0,0 +1,168 @@
+#!/usr/bin/python
+# Copyright (c) 2017 Dell Inc.
```
Comment (atanu-mandal) [inline scripts/bin/base_nas_mac_config.py]: Copyright update?

File: scripts/bin/base_nas_fanout_init.sh [hunk 10]
```diff
@@ -0,0 +1,21 @@
+#!/bin/bash
```
Comment (atanu-mandal) [inline scripts/bin/base_nas_fanout_init.sh]: This file is removed in https://github.com/open-switch/opx-nas-interface/commit/78168e4e69fc99ca635dea9962d453a886d1e94c . Please don't bring this in.

File: scripts/bin/base_nas_fanout_init.py [hunk 11]
```diff
@@ -18,13 +18,11 @@
 import event_log as ev
```
Comment (atanu-mandal) [inline scripts/bin/base_nas_fanout_init.py]: Revert the changes in this file. This would break https://github.com/open-switch/opx-nas-interface/commit/78168e4e69fc99ca635dea9962d453a886d1e94c

File: scripts/init/opx-create-interface.service [hunk 12]
```diff
@@ -1,16 +1,16 @@
...
+Requires=opx-cps.service opx-nas.service
 
 [Service]
 Type=oneshot
 RemainAfterExit=yes
 EnvironmentFile=/etc/opx/opx-environment
 ExecStart=/usr/bin/python -u /usr/bin/base_nas_create_interface.py
-ExecStartPost=/usr/bin/python -u /usr/bin/base_nas_fanout_init.py && "/bin/sh -c /usr/bin/network_restart.sh"
-TimeoutStartSec=180
+ExecStartPost=/bin/sh -c "/bin/sh -c /usr/bin/base_nas_fanout_init.sh && /usr/bin/network_restart.sh"
```
Comment (atanu-mandal) [inline scripts/init/opx-create-interface.service]: Revert this change. Refer https://github.com/open-switch/opx-nas-interface/commit/78168e4e69fc99ca635dea9962d453a886d1e94c

File: scripts/lib/python/nas_bridge_config_obj.py [hunk 13]
```diff
@@ -0,0 +1,134 @@
+#!/usr/bin/python
+# Copyright (c) 2017 Dell Inc.
```
Comment (atanu-mandal) [inline scripts/lib/python/nas_bridge_config_obj.py]: Copyright update?

File: scripts/lib/python/nas_if_bridge.py [hunk 14]
```diff
@@ -0,0 +1,213 @@
+#!/usr/bin/python
+# Copyright (c) 2017 Dell Inc.
```
Comment (atanu-mandal) [inline scripts/lib/python/nas_if_bridge.py]: copyright update?

File: scripts/lib/python/nas_if_handler_lib.py [hunk 15]
```diff
@@ -0,0 +1,282 @@
+#!/usr/bin/python
+# Copyright (c) 2015 Dell Inc.
```
Comment (atanu-mandal) [inline scripts/lib/python/nas_if_handler_lib.py]: copyright update?

File: scripts/lib/python/nas_if_vlan_subintf.py [hunk 16]
```diff
@@ -0,0 +1,108 @@
+#!/usr/bin/python
+# Copyright (c) 2017 Dell Inc.
```
Comment (atanu-mandal) [inline scripts/lib/python/nas_if_vlan_subintf.py]: copyright update?

File: scripts/lib/python/nas_if_vtep.py [hunk 17]
```diff
@@ -0,0 +1,184 @@
+#!/usr/bin/python
+# Copyright (c) 2017 Dell Inc.
```
Comment (atanu-mandal) [inline scripts/lib/python/nas_if_vtep.py]: copyright update?

File: scripts/lib/python/nas_vlan_subintf_config_obj.py [hunk 18]
```diff
@@ -0,0 +1,90 @@
+#!/usr/bin/python
+# Copyright (c) 2017 Dell Inc.
```
Comment (atanu-mandal) [inline scripts/lib/python/nas_vlan_subintf_config_obj.py]: copyright update?

File: scripts/lib/python/nas_vtep_config_obj.py [hunk 19]
```diff
@@ -0,0 +1,193 @@
+#!/usr/bin/python
+# Copyright (c) 2017 Dell Inc.
```
Comment (atanu-mandal) [inline scripts/lib/python/nas_vtep_config_obj.py]: copyright update?

File: scripts/lib/python/nas_vxlan_utils.py [hunk 20]
```diff
@@ -0,0 +1,208 @@
+#!/usr/bin/python
+# Copyright (c) 2017 Dell Inc.
```
Comment (atanu-mandal) [inline scripts/lib/python/nas_vxlan_utils.py]: copyright update?

File: src/bridge/nas_vlan_bridge_cps.cpp [hunk 21]
```diff
@@ -0,0 +1,651 @@
+/*
+ * Copyright (c) 2016 Dell Inc.
```
Comment (atanu-mandal) [inline src/bridge/nas_vlan_bridge_cps.cpp]: copyright update?

File: src/unit_test/nas_int_vxlan_unittest.cpp [hunk 22]
```diff
@@ -0,0 +1,218 @@
+/*
+ * Copyright (c) 2016 Dell Inc.
```
Comment (atanu-mandal) [inline src/unit_test/nas_int_vxlan_unittest.cpp]: copyright update?

File: src/vrf/nas_vrf_api.cpp [hunk 23]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2018 Dell Inc.
+ * Copyright (c) 2017 Dell Inc.
```
Comment (atanu-mandal) [inline src/vrf/nas_vrf_api.cpp]: Revert this.

**Comment by GarrickHe** [comment]
Pull-request updated, HEAD is now d715d24f3e0bd0948a923150e4d479c512ef9568

**Comment by GarrickHe** [comment]
Pull-request updated, HEAD is now aa380efbc865da1f25b6e74149f331d2d162aff3

**Comment by GarrickHe** [comment]
Pull-request updated, HEAD is now 7fcf54504333f284ba8c28bebfda61cb2a9d911c

File: scripts/bin/base_nas_fanout_init.sh (hunk 10 above)
Comment (atanu-mandal) [reply scripts/bin/base_nas_fanout_init.sh]: Please address this.

File: scripts/bin/base_nas_fanout_init.py (hunk 11 above)
Comment (atanu-mandal) [reply scripts/bin/base_nas_fanout_init.py]: Please address this.

File: scripts/init/opx-create-interface.service (hunk 12 above)
Comment (atanu-mandal) [reply scripts/init/opx-create-interface.service]: Please address this.

**Comment by GarrickHe** [comment]
Pull-request updated, HEAD is now 71eaa6f346561445185bf54b7b4ffca6e22ef4d0

**Comment by GarrickHe** [comment]
Pull-request updated, HEAD is now 71eaa6f346561445185bf54b7b4ffca6e22ef4d0

**Comment by jeff-yin** [comment]
Admin-merged this due to a buildkite timeout issue.
